Fixes #26783 - Create Statistics Route

glekner opened the pull request:
<!--
Simple description of what is fixed/introduced.

Prerequisites for testing:
* VMware cluster
* Two smart proxies

Tests needed (anticipated impacts):
* Hosts list - mainly power column
* Power status on host page
-->

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
* Suggest prerequisites for testing and testing scenarios following example above.
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

ohadlevy commented on webpack/assets/javascripts/react_app/routes/Statistics/StatisticsPage/StatisticsPage.js:
```diff
@@ -0,0 +1,43 @@
+
+class Statistics extends Component {
+  componentDidMount() {
+    document.title = __('Statistics');
```
should this be part of the PageLayout component?

ohadlevy commented on webpack/assets/javascripts/react_app/routes/Statistics/constants.js:
```diff
@@ -0,0 +1,5 @@
+export const STATISTICS_PAGE_HIDE_LOADING = 'STATISTICS_PAGE_HIDE_LOADING';
+export const STATISTICS_PAGE_SHOW_MESSAGE = 'STATISTICS_PAGE_SHOW_MESSAGE';
+
+export const STATISTICS_PAGE_URL = '/statistics';
```
can we use the foremanUrl function ?

ohadlevy commented on webpack/assets/javascripts/react_app/routes/common/EmptyPage/EmptyPage.js:
```diff
@@ -0,0 +1,36 @@
+const EmptyPage = ({ type, text, isLoading }) => (
+  <React.Fragment>
+    {isLoading ? (
+      <div id="loading-page">
```
can you use the LoadingState component instead?

glekner replied:
we can't customize `loadingState` with different skeletons like we talked about, also the component only takes care of the loading, what about empty results or a network error?
I wrote `EmptyPage` that shows a custom loading component and empty/error result if necessary.

ohadlevy replied:
In my mind, they are not connected, e.g. everything that has a loading timeout, should be wrapped in a LoadingState component, what you show, is irrelevant to the loading component ? 

not sure I understand why you went this path of recreating parts of it?

glekner replied:
the `loadingState` can't be customized, it always shows a spinner, we want to customize the loading view per view. I tried to create a component which takes care of the loading and also of the empty/error states called EmptyPage, but we can also separate them.

sharvit commented on app/controllers/react_controller.rb:
```diff
@@ -1,3 +1,7 @@
 end
+
+def index
+  render 'react/index'
```
Instead of creating an empty view, is it possible to render the `layouts/react_application` without rendering a view inside?
@ohadlevy @tbrisker @ShimShtein?

sharvit commented on webpack/assets/javascripts/react_app/routes/Statistics/StatisticsPage/StatisticsPageActions.js:
```diff
@@ -0,0 +1,30 @@
+
+export const fetchStatistics = (url = STATISTICS_PAGE_URL) => dispatch => {
+  const onFetchSuccess = ({ data }) => {
+    dispatch({ type: STATISTICS_PAGE_HIDE_LOADING });
```
IMO `hideLoading` should be an action

sharvit commented on webpack/assets/javascripts/react_app/routes/Statistics/StatisticsPage/StatisticsPageActions.js:
```diff
@@ -0,0 +1,30 @@
+  STATISTICS_PAGE_URL,
+} from '../constants';
+
+export const fetchStatistics = (url = STATISTICS_PAGE_URL) => dispatch => {
```
I believe it would look better with `async/await`

sharvit reviewed (commented):
Thanks @glekner :+1:

glekner commented:
@ohadlevy looks like `foremanUrl` is [not defined ](https://travis-ci.org/theforeman/foreman/jobs/531420179#L1733)in tests

ohadlevy commented:
> @ohadlevy looks like `foremanUrl` is [not defined ](https://travis-ci.org/theforeman/foreman/jobs/531420179#L1733)in tests


of course - see https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/foreman_tools.js#L131

glekner commented:
- added `URL_PREFIX` to jest's globals, it fixed 1 test but the rest fails with `TypeError: (0 , _foreman_tools.foremanUrl) is not a function`
- added @sharvit suggestions

glekner commented:
New suggestion, use HOCs to handle the render state (Loading, Error, Empty), get rid of class components, and make the code more readable.

I created `WithRenderHandler` HOC

```js
const WithRenderHandler = (
  Component,
  LoadingComponent = LoadingPage,
  EmptyComponent = EmptyPage,
  ErrorComponent = ErrorPage,
) => componentProps => {
  const { isLoading, hasData, hasError, message } = componentProps; // must have props

  if (isLoading) return <LoadingComponent />;
  if (hasError) return <ErrorComponent />;
  if (!hasData) return <EmptyComponent />;
  return <Component {...componentProps} />;
};
```

This takes care of the conditional rendering inside the component, but what about lifecycle methods? we can use `lifecycle` from `recompose`, but for the concept I created a helper that runs a `componentDidMount`:

```js
const callOnMount = callback => WrappedComponent =>
  class extends React.Component {
    componentDidMount() {
      callback(this.props);
    }

    render() {
      return <WrappedComponent {...this.props} />;
    }
  };
```
Now lets take Statistics view for example, here is the updated `index.js`:

```js
export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  callOnMount(({ fetchStatistics }) => fetchStatistics())
)(StatisticsPage);
```
and here is the updated and **little** `StatisticsPage.js`:

```js
const Statistics = ({ statistics }) => (
  <PageLayout header={__('Statistics')} searchable={false}>
    <StatisticsChartsList data={statistics} />
  </PageLayout>
);

export default WithRenderHandler(Statistics, StatisticsLoadingPage); // we pass only a custom loading component
```

Thoughts @sharvit @ohadlevy @amirfefer ?

ohadlevy commented on package.json:
```diff
@@ -159,7 +159,8 @@
     "globals": {
-      "__testing__": true
+      "__testing__": true,
+      "URL_PREFIX": "/"
```
why?

glekner replied:
to mock `URL_PREFIX` in tests

ohadlevy commented on webpack/assets/javascripts/react_app/routes/Statistics/StatisticsPage/StatisticsLoadingPage/statisticsloadingpage.scss:
```diff
@@ -0,0 +1,16 @@
+  display: flex;
+  flex-wrap: wrap;
+
+  #empty-statistics {
```
maybe we should call it loader or something that its clear its a temporary loader class? eg empty has another meaning?

ohadlevy commented on webpack/assets/javascripts/react_app/routes/Statistics/StatisticsPage/StatisticsPageActions.js:
```diff
@@ -0,0 +1,44 @@
+    dispatch(hideLoading());
+    dispatch({
+      type: STATISTICS_PAGE_FETCH_META,
+      payload: { statistics: data, hasData: data.length > 0 },
```
shouldn't hasData be a selector?

ohadlevy commented on webpack/assets/javascripts/react_app/routes/Statistics/constants.js:
```diff
@@ -0,0 +1,9 @@
+
+export const STATISTICS_PAGE_URL = foremanUrl
+  ? foremanUrl('/statistics')
+  : '/statistics';
```
imho this is not required, the function will return the right value

glekner replied:
@sharvit `foremanUrl` is undefined in tests for some reason. any idea?

ohadlevy replied:
is this still an issue with the mock?

glekner replied:
yes, `foremanUrl` is undefined in tests.

ohadlevy replied:
hmm.. any idea why? @sharvit @amirfefer ?

ohadlevy commented on webpack/assets/javascripts/react_app/routes/common/EmptyPage/index.js:
```diff
@@ -0,0 +1,27 @@
+import DefaultEmptyState from '../../../components/common/EmptyState';
+import './emptypage.scss';
+
+const EmptyPage = ({ type, text, name }) => (
```
whats the difference between empty page and empty state?

glekner replied:
no difference, just a default one with i18n

ohadlevy replied:
can't we add the i18n? I think we already have a wrapper in foreman for it?

ohadlevy commented on webpack/assets/javascripts/react_app/routes/common/LoadingPage/index.js:
```diff
@@ -0,0 +1,11 @@
+
+const LoadingPage = () => (
+  <div id="loading-page">
+    <Spinner size="lg" loading />
```
why not use loadingState here?

glekner replied:
loadingState needs children and handles which component to render if i'm not mistaken, the idea of this file is just to be a "stupid" default skeleton

ohadlevy replied:
I still like the fact that you can show the spinner after a certain amount of ms and not to show it if its too fast, I wonder if can't just use it without childrens?

glekner replied:
I added `loadingState` with 500ms delay for the default Loading page, its great for spinners.
For custom skeletons, I think its fine to show them right away.

ohadlevy reviewed (changes requested):
LGTM with minor nitpicks, but looking forward to @sharvit and @amirfefer review for the mechanism... thanks!

sharvit commented:
> New suggestion, use HOCs to handle the render state (Loading, Error, Empty), get rid of class components, and make the code more readable.
> 
> I created `WithRenderHandler` HOC
> 
> ```js
> const WithRenderHandler = (
>   Component,
>   LoadingComponent = LoadingPage,
>   EmptyComponent = EmptyPage,
>   ErrorComponent = ErrorPage,
> ) => componentProps => {
>   const { isLoading, hasData, hasError, message } = componentProps; // must have props
> 
>   if (isLoading) return <LoadingComponent />;
>   if (hasError) return <ErrorComponent />;
>   if (!hasData) return <EmptyComponent />;
>   return <Component {...componentProps} />;
> };
> ```
> 
> This takes care of the conditional rendering inside the component, but what about lifecycle methods? we can use `lifecycle` from `recompose`, but for the concept I created a helper that runs a `componentDidMount`:
> 
> ```js
> const callOnMount = callback => WrappedComponent =>
>   class extends React.Component {
>     componentDidMount() {
>       callback(this.props);
>     }
> 
>     render() {
>       return <WrappedComponent {...this.props} />;
>     }
>   };
> ```
> 
> Now lets take Statistics view for example, here is the updated `index.js`:
> 
> ```js
> export default compose(
>   connect(
>     mapStateToProps,
>     mapDispatchToProps
>   ),
>   callOnMount(({ fetchStatistics }) => fetchStatistics())
> )(StatisticsPage);
> ```
> 
> and here is the updated and **little** `StatisticsPage.js`:
> 
> ```js
> const Statistics = ({ statistics }) => (
>   <PageLayout header={__('Statistics')} searchable={false}>
>     <StatisticsChartsList data={statistics} />
>   </PageLayout>
> );
> 
> export default WithRenderHandler(Statistics, StatisticsLoadingPage); // we pass only a custom loading component
> ```
> 
> Thoughts @sharvit @ohadlevy @amirfefer ?

I love it and got some suggestions

1. IMO the `WithRenderHandler` can run the `onMount` internally by returning a class or by using `useEffect`. It will reduce the effort you need in order to create a page.

2. It would be nice to be able to reuse reducers as well as they are going to be similar whenever we will want to use the `WithRenderHandler` or outside of the `WithRenderHandler`.
See how the Table reuse reducers: 
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/common/table/reducers/createTableReducer.js
https://github.com/theforeman/foreman/blob/2472a9d40e9887a5b6eef5ef4338825430ec3b8e/webpack/assets/javascripts/react_app/components/ModelsTable/ModelsTableReducer.js

sharvit commented on webpack/assets/javascripts/react_app/routes/Statistics/StatisticsPage/StatisticsPage.js:
```diff
@@ -0,0 +1,18 @@
+import PageLayout from '../../../pages/common/PageLayout/PageLayout';
+import StatisticsChartsList from '../../../components/statistics/StatisticsChartsList';
+
+const Statistics = ({ statistics }) => (
```
nitpick: `const StatisticsPage`

sharvit commented on webpack/assets/javascripts/react_app/routes/Statistics/StatisticsPage/StatisticsPage.js:
```diff
@@ -0,0 +1,18 @@
+  statistics: PropTypes.array.isRequired,
+};
+
+export default WithRenderHandler(Statistics, StatisticsLoadingPage);
```
IMO it should be called in the integration file (`index.js`)

sharvit commented on webpack/assets/javascripts/react_app/routes/Statistics/StatisticsPage/StatisticsPageReducer.js:
```diff
@@ -0,0 +1,31 @@
+  const { payload } = action;
+
+  switch (action.type) {
+    case STATISTICS_PAGE_FETCH_META:
```
Why `meta` isn't it the actual statistics data?

glekner replied:
its meta, the component inside the page actually fetches the statistics

sharvit commented on webpack/assets/javascripts/react_app/routes/Statistics/StatisticsPage/StatisticsPageReducer.js:
```diff
@@ -0,0 +1,31 @@
+  message: {},
+});
+
+export default (state = initialState, action) => {
```
Shouldn't this reducer use the async convention of?
`STATISTICS_PAGE_FETCH_REQUEST`
`STATISTICS_PAGE_FETCH_SUCCESS`
`STATISTICS_PAGE_FETCH_FAILED`

sharvit commented on webpack/assets/javascripts/react_app/routes/common/WithRenderHandler/index.js:
```diff
@@ -0,0 +1,18 @@
+
+  if (isLoading) return <LoadingComponent />;
+  if (hasError) return <EmptyComponent type="error" text={message.text} />;
+  if (!hasData) return <EmptyComponent type="empty" />;
```
IMO `EmptyComponent` and `ErrorComponent` should be different components.

sharvit commented on webpack/assets/javascripts/react_app/routes/common/WithRenderHandler/index.js:
```diff
@@ -0,0 +1,18 @@
+  Component,
+  LoadingComponent = LoadingPage,
+  EmptyComponent = EmptyPage
+) => componentProps => {
```
It would be nice to use:
```js
const WithRenderHandler = ({
  Component,
  LoadingComponent = LoadingPage,
  EmptyComponent = EmptyPage
})
```

So I can only supply what I want to override `WithRenderHandler({ Component, EmptyComponent  })`

sharvit commented on webpack/assets/javascripts/react_app/routes/common/WithRenderHandler/index.js:
```diff
@@ -0,0 +1,18 @@
+import EmptyPage from '../EmptyPage';
+import LoadingPage from '../LoadingPage';
+
+const WithRenderHandler = (
```
nitpick: the convention for HOCs are `withRenderHandler`

sharvit replied:
The name `withRenderHandler` is confusing and I don't feel it tell me about what it does.
What about `withManagedData`?

glekner replied:
it handles which component to render (loading, empty etc..) its not part of the API hoc

sharvit commented on webpack/assets/javascripts/react_app/routes/common/callOnMount/index.js:
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+
+const callOnMount = callback => WrappedComponent =>
```
IMO the `withRenderHandler` can run `componentDidMount` internally.
I believe once it has done, this HOC will not be needed anymore.

I don't see a benefit using it instead `useEffect` or `componentDidMount`. I actually think it takes for developers more time to understand how things works when using HOCs as they can be very confusing.

glekner replied:
we dont need this, we can use [lifecycle](https://github.com/acdlite/recompose/blob/master/docs/API.md#lifecycle), I wrote this for things to be more understandable,

sharvit reviewed (changes requested):
Thanks @glekner 👍

glekner commented:
- `StatisticsChartList` is no longer connected
- Page now manages the charts state to avoid duplication.
- removed `statistics` folder from reducers, actions in the legacy structure
- `callOnMount` and `withRenderHandler` HOCs are used 🚀

glekner committed:
Fixes #26783 - Create Statistics Route

glekner commented:
- Reverted Redux refactoring.

This new Route now uses all the recent infrastructure additions `withRenderHandler, callOnMount, withDataReducer`. @amirfefer @tbrisker @xprazak2  This is a good first example

Only 360 Lines including snapshots, can we merge @ohadlevy :) ?

glekner commented on webpack/assets/javascripts/react_app/routes/Statistics/StatisticsPage/StatisticsPageActions.js:
```diff
@@ -0,0 +1,44 @@
+} from '../constants';
+
+export const getStatisticsMeta = (
+  url = foremanUrl(STATISTICS_PAGE_URL)
```
moving `foremanUrl` to the actions file fixed the tests @ohadlevy

glekner commented:
Are there additional comments @ohadlevy ?
Test failures unrelated
